course-material/interactive-queries-and-parameterization/parameterize-sql-queries.md

```diff
@@ -1,14 +1,358 @@
+
+    # Save the cleaned up CSV file
+    df.to_csv(df.to_csv(f"{data_name}_cleaned.csv", index=False))
+
+
+# Running the above function
+extract_to_csv("https://tinyurl.com/uci-marketing-data", "bank")
+```
+
+Initialize a DuckDB Instance
+
+```{code-cell} ipython3
+# Loading in SQL extension
+%reload_ext sql
+# Initiating a DuckDB database named 'bank.duck.db' to run our SQL queries on
+%sql duckdb:///bank.duck.db
+```
+
+Load the data
+
+```{code-cell} ipython3
+%%sql
+CREATE OR REPLACE TABLE bank AS
+FROM read_csv_auto('bank_cleaned.csv', header=True, sep=',')
+```
+
+We confirm the table was loaded
+
+```{code-cell} ipython3
+%sqlcmd tables
+```
+
+We can use [JupySQL's Table Explorer](https://jupysql.ploomber.io/en/latest/user-guide/table_explorer.html) to take a quick look at the table.
+
+```{code-cell} ipython3
+%sqlcmd explore --table bank
+```
+
+```{code-cell} ipython3
+%sqlcmd columns -t bank
+```
+
+## Variable Expansion
+
+JupySQL supports variable expansion in the form of {{variable}}. This allows you to write a query with placeholders that can be replaced by variables dynamically. The benefits of using parameterized SQL queries are that they can be reused with different values, prepared ahead of time, and used with dynamic data.
+
+Let's start with a simple query template with placeholders, and substitute the placeholders with a couple of variables using variable expansion.
+
+We are going to select two columns (`age` and `job`) from the table `bank` with the variable `dynamic_column` and show only 5 rows. We can control this via another variable `dynamic_limit`.
+
+```{code-cell} ipython3
+dynamic_limit = 5
+dynamic_column = "age, job"
+%sql SELECT {{dynamic_column}} FROM bank LIMIT {{dynamic_limit}}
+```
+
+### Sample case
+
+The HR department would like to determine top five oldest and youngest employees in two different jobs:  services, management.
+
+We can combine Python with SQL code using the `%sql` magic. With Python we can iterate over the jobs in the list `jobs`. With parameterization, we can
+then insert the job into the SQL query via the `{{}}` notation.
+
+```{code-cell} ipython3
+jobs = ["services", "management"]
+
+for job in jobs:
+    print(f"Top 5 oldest in {job}")
+    order = "DESC"
+    top_oldest = %sql SELECT * FROM bank WHERE job='{{job}}' ORDER BY age {{order}} LIMIT 5
+    print(top_oldest)
+
+    print(f"Top 5 youngest in {job}")
+    order = "ASC"
+    top_youngest = %sql SELECT * FROM bank WHERE job='{{job}}' ORDER BY age {{order}} LIMIT 5
+    print(top_youngest)
+```
+
+### Mini Exercise:
+
+The marketing team is interested in understanding the distribution of age for their marketing campaigns. Select the top 5 oldest and youngest people who have been part of the marketing campaign (`campaign` column).
+
+Your task: use a loop with variable expansion to accomplish this. 
+
+<!-- #region -->
+<details>
+
+<summary>Answers</summary>
+
+```{code-cell} ipython3
+campaigns = [1, 2, 3, 4, 5]
```

Comment:
Not sure if it's necessary: because the unique values for `campaigns` are hard-coded, maybe we could use pandas (data already loaded in `df`) to make the list? My main concern is that for some columns, hard-coding values might not be efficient and this helps address as many use cases not just our example